### PR TITLE
Remove Client, Request, Response, TypedInput, and TypedOutput abstractions.

### PR DESCRIPTION
**Pull request by JakeWharton**
This disables multipart and form URL encoding for now. Like logging, it shall return!

### CODE REVIEW

**Comment by JakeWharton** [comment]
@swankjesse biggun'

**Comment by JakeWharton** [comment]
After this commit Retrofit's jar is 49.5% smaller than the last release. More weight watchers to come.

File: retrofit-converters/protobuf/src/main/java/retrofit/converter/ProtoConverter.java [hunk 1]
```diff
@@ -24,30 +24,31 @@
       throw new IllegalArgumentException("Expected a protobuf message but was " + c.getName());
     }
 
-    String mimeType = body.mimeType();
```
Comment (swankjesse) [inline retrofit-converters/protobuf/src/main/java/retrofit/converter/ProtoConverter.java]: Why drop this?

File: retrofit-converters/simplexml/src/main/java/retrofit/converter/SimpleXMLConverter.java [hunk 2]
```diff
@@ -20,7 +19,8 @@
 public class SimpleXMLConverter implements Converter {
   private static final boolean DEFAULT_STRICT = true;
   private static final String CHARSET = "UTF-8";
-  private static final String MIME_TYPE = "application/xml; charset=" + CHARSET;
+  private static final MediaType MEDIA_TYPE =
+      MediaType.parse("application/xml; charset=" + CHARSET);
```
Comment (swankjesse) [inline retrofit-converters/simplexml/src/main/java/retrofit/converter/SimpleXMLConverter.java]: nit: typically `UTF-8` is lowercase in a content type.

Comment (swankjesse) [inline retrofit-converters/simplexml/src/main/java/retrofit/converter/SimpleXMLConverter.java]: (though I have no evidence to back this up!

File: retrofit-converters/simplexml/src/main/java/retrofit/converter/SimpleXMLConverter.java [hunk 3]
```diff
@@ -43,35 +43,34 @@ public SimpleXMLConverter(Serializer serializer, boolean strict) {
...
     }
   }
 
-  @Override public TypedOutput toBody(Object source, Type type) {
-    OutputStreamWriter osw = null;
-
+  @Override public RequestBody toBody(Object source, Type type) {
+    byte[] bytes;
     try {
       ByteArrayOutputStream bos = new ByteArrayOutputStream();
```
Comment (swankjesse) [inline retrofit-converters/simplexml/src/main/java/retrofit/converter/SimpleXMLConverter.java]: Use a Buffer instead?

File: retrofit-converters/simplexml/src/main/java/retrofit/converter/SimpleXMLConverter.java [hunk 4]
```diff
@@ -43,35 +43,34 @@ public SimpleXMLConverter(Serializer serializer, boolean strict) {
     this.strict = strict;
   }
 
-  @Override public Object fromBody(TypedInput body, Type type) throws IOException {
+  @Override public Object fromBody(ResponseBody body, Type type) throws IOException {
+    InputStream is = body.byteStream();
     try {
-      return serializer.read((Class<?>) type, body.in(), strict);
+      return serializer.read((Class<?>) type, is, strict);
```
Comment (swankjesse) [inline retrofit-converters/simplexml/src/main/java/retrofit/converter/SimpleXMLConverter.java]: Tempting to suggest calling `ResponseBody.charStream()` instead of `ResponseBody.byteStream()`, but it's a potentially-breaking behavior change. XML documents that don't specify the charset in the content-type header can instead specify it in the `<?xml` declaration thingy.

Leave this as-is.

File: retrofit-converters/simplexml/src/test/java/retrofit/converter/SimpleXMLConverterTest.java [hunk 5]
```diff
@@ -9,65 +15,55 @@
...
 import org.simpleframework.xml.stream.Verbosity;
-import retrofit.mime.TypedByteArray;
-import retrofit.mime.TypedInput;
-import retrofit.mime.TypedOutput;
 
 import static org.assertj.core.api.Assertions.assertThat;
 
 public class SimpleXMLConverterTest {
-	private static final String MIME_TYPE = "application/xml; charset=UTF-8";
+	private static final MediaType MEDIA_TYPE = MediaType.parse("application/xml; charset=UTF-8");
```
Comment (swankjesse) [inline retrofit-converters/simplexml/src/test/java/retrofit/converter/SimpleXMLConverterTest.java]: +6 ??

Comment (swankjesse) [inline retrofit-converters/simplexml/src/test/java/retrofit/converter/SimpleXMLConverterTest.java]: (indentation)

File: retrofit-converters/simplexml/src/test/java/retrofit/converter/SimpleXMLConverterTest.java [hunk 6]
```diff
@@ -9,65 +15,55 @@
...
 
-	@Test
-	public void deserialize() throws Exception {
-		final TypedInput input = new TypedByteArray(MIME_TYPE,
-				objAsXML.getBytes());
-		final MyObject result = (MyObject) converter.fromBody(input,
-				MyObject.class);
-		assertThat(result).isEqualTo(obj);
+	@Test public void deserialize() throws Exception {
+    ResponseBody body = ResponseBody.create(MEDIA_TYPE, XML);
```
Comment (swankjesse) [inline retrofit-converters/simplexml/src/test/java/retrofit/converter/SimpleXMLConverterTest.java]: ... but sometimes +2

File: retrofit-converters/wire/src/main/java/retrofit/converter/WireConverter.java [hunk 7]
```diff
@@ -36,31 +36,24 @@ public WireConverter(Wire wire) {
       throw new IllegalArgumentException("Expected a proto message but was " + c.getName());
     }
 
-    if (!MIME_TYPE.equalsIgnoreCase(body.mimeType())) {
-      throw new IllegalStateException("Expected a proto but was: " + body.mimeType());
```
Comment (swankjesse) [inline retrofit-converters/wire/src/main/java/retrofit/converter/WireConverter.java]: 2 things:
I don't mind dropping this check.
It should probably be a separate PR. (So it can be rolled-back or referenced independently).

Comment (swankjesse) [inline retrofit-converters/wire/src/main/java/retrofit/converter/WireConverter.java]: .... although here I think you're making so many forward-progress but incompatible changes that it doesn't really matter. You aren't working to a .1 release, this is a major release.

File: retrofit-mock/src/main/java/retrofit/MockResponseBody.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
+// Copyright 2013 Square, Inc.
```
Comment (swankjesse) [inline retrofit-mock/src/main/java/retrofit/MockResponseBody.java]: Big Apache copyright header with 2015 ?

Comment (swankjesse) [inline retrofit-mock/src/main/java/retrofit/MockResponseBody.java]: ... found the origin. Nevermind

File: retrofit/src/main/java/retrofit/ExceptionCatchingRequestBody.java [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    return delegate.contentType();
+  }
+
+  @Override public long contentLength() {
+    return delegate.contentLength();
+  }
+
+  @Override public BufferedSource source() {
+    return Okio.buffer(new ForwardingSource(delegate.source()) {
+      @Override public long read(Buffer sink, long byteCount) throws IOException {
```
Comment (swankjesse) [inline retrofit/src/main/java/retrofit/ExceptionCatchingRequestBody.java]: Do we care about close?

File: retrofit/src/main/java/retrofit/RestAdapter.java [hunk 10]
```diff
@@ -183,69 +180,46 @@ static RestMethodInfo getMethodInfo(Map<Method, RestMethodInfo> cache, Method me
...
-          error.set(e);
-          latch.countDown();
-        }
-      });
-      latch.await();
-
-      RetrofitError actualError = error.get();
-      if (actualError != null) {
-        throw handleError(actualError);
+    private Object invokeSync(MethodInfo methodInfo, Request request) throws Throwable {
```
Comment (swankjesse) [inline retrofit/src/main/java/retrofit/RestAdapter.java]: Love this.

Comment (swankjesse) [inline retrofit/src/main/java/retrofit/RestAdapter.java]: Fewer idle threads!

File: retrofit/src/main/java/retrofit/RestAdapter.java [hunk 11]
```diff
@@ -254,65 +228,84 @@ private Object invokeRx(final RestMethodInfo methodInfo, final Request request)
...
       }
 
-      ExceptionCatchingTypedInput wrapped = new ExceptionCatchingTypedInput(body);
+      ExceptionCatchingRequestBody wrapped = new ExceptionCatchingRequestBody(body);
       try {
-        Object convert = converter.fromBody(wrapped, type);
-        callResponse(callback, convert, response);
+        return converter.fromBody(wrapped, type);
       } catch (RuntimeException e) {
         // If the underlying input stream threw an exception, propagate that rather than
```
Comment (swankjesse) [inline retrofit/src/main/java/retrofit/RestAdapter.java]: This behavior is really nice.

File: retrofit/src/main/java/retrofit/RetrofitError.java [hunk 12]
```diff
@@ -15,28 +15,34 @@
...
     return new RetrofitError(exception.getMessage(), url, null, null, null, Kind.NETWORK,
         exception);
   }
 
-  public static RetrofitError unexpectedError(String url, Response response, Converter converter,
-      Type successType, Throwable exception) {
-    return new RetrofitError(exception.getMessage(), url, response, converter, successType,
-        Kind.UNEXPECTED, exception);
+  public static RetrofitError networkError(Response response, IOException exception) {
+    response = response.newBuilder().body(null).build(); // Remove any body.
```
Comment (swankjesse) [inline retrofit/src/main/java/retrofit/RetrofitError.java]: Nice.

File: retrofit/src/main/java/retrofit/RetrofitError.java [hunk 13]
```diff
@@ -117,14 +123,14 @@ public Object getBodyAs(Type type) {
...
-    TypedInput body = response.getBody();
+    ResponseBody body = response.body();
     if (body == null) {
       return null;
     }
     try {
       return converter.fromBody(body, type);
     } catch (IOException e) {
-      throw new RuntimeException(e); // Body is a byte[], can't be a real IO exception.
+      throw new RuntimeException(e); // Body is a Buffer, can't be a real IO exception.
```
Comment (swankjesse) [inline retrofit/src/main/java/retrofit/RetrofitError.java]: ... sometimes we use `ProtocolException` to indicate malformed input.

File: retrofit/src/main/java/retrofit/Utils.java [hunk 14]
```diff
@@ -35,53 +34,35 @@
...
+
+          @Override public long contentLength() {
+            return buffer.size();
+          }
 
-  static Response replaceResponseBody(Response response, TypedInput body) {
-    return new Response(response.getUrl(), response.getStatus(), response.getReason(),
-        response.getHeaders(), body);
+          @Override public BufferedSource source() {
+            return buffer;
```
Comment (swankjesse) [inline retrofit/src/main/java/retrofit/Utils.java]: Do we need to clone things here?

Comment (swankjesse) [inline retrofit/src/main/java/retrofit/Utils.java]: I think I want an API to take a cheap, efficient snapshot of a buffer, which can be used as a factory for sources. . . . 

Buffer originalMutable = ...
ByteString snapshotImmutable = buffer.snapshot(); // Doesn't consume buffer!
BufferedSource copyReader = snapshotImmutable.source();

We could do this efficiently.

Comment (swankjesse) [inline retrofit/src/main/java/retrofit/Utils.java]: ... though I don't know if the result of `.snapshot` should be a ByteString that's backed by immutable segments, or if it should be a new type . . . _probably_ it should be a ByteString, so we have exactly one type to model an immutable sequence of bytes.

File: retrofit/src/main/java/retrofit/converter/GsonConverter.java [hunk 15]
```diff
@@ -45,69 +46,38 @@ public GsonConverter() {
...
 
   /**
    * Create an instance using the supplied {@link Gson} object for conversion. Encoding to JSON and
    * decoding from JSON (when no charset is specified by a header) will use the specified charset.
    */
-  public GsonConverter(Gson gson, String charset) {
+  public GsonConverter(Gson gson, Charset charset) {
     this.gson = gson;
     this.charset = charset;
+    this.mediaType = MediaType.parse("application/json; charset=" + charset.name());
```
Comment (swankjesse) [inline retrofit/src/main/java/retrofit/converter/GsonConverter.java]: Do we want an API like this?

```
  MediaType.parse("application/json").withCharset(charset)
```

File: retrofit/src/main/java/retrofit/converter/GsonConverter.java [hunk 16]
```diff
@@ -45,69 +46,38 @@ public GsonConverter() {
...
   }
 
   /**
    * Create an instance using the supplied {@link Gson} object for conversion. Encoding to JSON and
    * decoding from JSON (when no charset is specified by a header) will use the specified charset.
    */
-  public GsonConverter(Gson gson, String charset) {
+  public GsonConverter(Gson gson, Charset charset) {
     this.gson = gson;
     this.charset = charset;
```
Comment (swankjesse) [inline retrofit/src/main/java/retrofit/converter/GsonConverter.java]: check not null ?

File: retrofit/src/main/java/retrofit/converter/GsonConverter.java [hunk 17]
```diff
@@ -45,69 +46,38 @@ public GsonConverter() {
...
+    if (body.contentType() != null) {
+      charset = body.contentType().charset(charset);
     }
-    InputStreamReader isr = null;
+
+    InputStream is = body.byteStream();
     try {
-      isr = new InputStreamReader(body.in(), charset);
-      return gson.fromJson(isr, type);
+      return gson.fromJson(new InputStreamReader(is, charset), type);
```
Comment (swankjesse) [inline retrofit/src/main/java/retrofit/converter/GsonConverter.java]: Should OkHttp have an overload for `body.charStream` that takes a default charset? We should.

File: retrofit/src/test/java/retrofit/MockClient.java [hunk 18]
```diff
@@ -0,0 +1,49 @@
+package retrofit;
```
Comment (swankjesse) [inline retrofit/src/test/java/retrofit/MockClient.java]: nit: copyright

File: retrofit/src/test/java/retrofit/MockClient.java [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+import com.squareup.okhttp.Interceptor;
+import com.squareup.okhttp.Request;
+import com.squareup.okhttp.Response;
+import java.io.IOException;
+import java.util.ArrayDeque;
+import java.util.Deque;
+
+import static com.squareup.okhttp.Protocol.HTTP_1_1;
+
+final class MockClient implements Interceptor {
```
Comment (swankjesse) [inline retrofit/src/test/java/retrofit/MockClient.java]: I like where this is going

**Comment by swankjesse** [comment]
# Laughing Gas Tickles Me

LGTM. Love this.

File: retrofit-converters/protobuf/src/main/java/retrofit/converter/ProtoConverter.java (hunk 1 above)
Comment (JakeWharton) [inline retrofit-converters/protobuf/src/main/java/retrofit/converter/ProtoConverter.java]: It's inconsistent with the other converters. We never check JSON or XML for them.

File: retrofit-converters/simplexml/src/test/java/retrofit/converter/SimpleXMLConverterTest.java (hunk 5 above)
Comment (JakeWharton) [inline retrofit-converters/simplexml/src/test/java/retrofit/converter/SimpleXMLConverterTest.java]: Yeah wow. Not sure what's happening here. Both of these files were contributed and maybe they contain tabs? Will reformat.

File: retrofit/src/main/java/retrofit/RetrofitError.java (hunk 13 above)
Comment (JakeWharton) [inline retrofit/src/main/java/retrofit/RetrofitError.java]: That's malformed at the HTTP transport layer, and it would have thrown by here (we consumed the entire body into a `Buffer`).

File: retrofit/src/main/java/retrofit/converter/GsonConverter.java (hunk 15 above)
Comment (JakeWharton) [inline retrofit/src/main/java/retrofit/converter/GsonConverter.java]: Something more granular would be nice. I'm building a string just to have it parsed and yanked apart.

File: retrofit-converters/wire/src/main/java/retrofit/converter/WireConverter.java (hunk 7 above)
Comment (JakeWharton) [inline retrofit-converters/wire/src/main/java/retrofit/converter/WireConverter.java]: This is machete season! Hack hack hack away.

File: retrofit/src/main/java/retrofit/ExceptionCatchingRequestBody.java (hunk 9 above)
Comment (JakeWharton) [inline retrofit/src/main/java/retrofit/ExceptionCatchingRequestBody.java]: No we do not, per you. If close throws we trap it.

File: retrofit-converters/wire/src/main/java/retrofit/converter/WireConverter.java (hunk 7 above)
Comment (swankjesse) [inline retrofit-converters/wire/src/main/java/retrofit/converter/WireConverter.java]: Thumbs up.

**Comment by casidiablo** [comment]
Wow... what's the reason for removing support for custom clients and/or client providers? A link to the thread where this was discussed works.

Thanks!

**Comment by JakeWharton** [comment]
There is no discussion thread. The abstractions serve no purpose now that OkHttp is all grown up. You'll still be able to use a custom client, but you'll use OkHttp's types as the request/response model.
